text/0048-session-keys-runtime-api.md

```diff
@@ -0,0 +1,102 @@
+# RFC-0048: Generate ownership proof for `SessionKeys`
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 13 November 2023                                                                    |
+| **Description** | Change `SessionKeys` runtime api to also create a proof of ownership for on chain registration. |
+| **Authors**     | Bastian Köcher                                                                    |
+
+## Summary
+
+When rotating/generating the `SessionKeys` of a node, the node calls into the runtime using the 
+`SessionKeys::generate_session_keys` runtime api. This runtime api function needs to be changed
+to add an extra parameter `owner` and to change the return value to also include the `proof` of 
+ownership. The `owner` should be the account id of the account setting the `SessionKeys` on chain 
+to allow the on chain logic the verification of the proof. The on chain logic is then able to proof 
+the possession of the private keys of the `SessionKeys` using the `proof`.
+
+## Motivation
+
+When a user sets new `SessionKeys` on chain the chain can currently not ensure that the user 
+actually has control over the private keys of the `SessionKeys`. With the RFC applied the chain is able 
+to ensure that the user actually is in possession of the private keys.
```

Comment:
Worded differently, you want to give the possibility to the implementation of `SessionKeys_generate_session_keys` needs to be able to know which `AccountId` the client controls?

Reply:
Yes. I mean `generate_session_keys` doesn't care as it just signs whatever you pass it.

Reply:
At least for me, the motivation does not go deep enough. Why should the chain care?

Reply:
> Why should the chain care?

Care that you are in possession of the keys? The chain doesn't really need to care. However, someone could may include a transaction faster than you with your public keys registered. So, someone could theoretically prevent that a validator is rotating its keys.

Reply:
LOL. Ok, makes sense! Thanks!